### PR TITLE
Remove redundant tests and wait for device to be disconnected

### PR DESCRIPTION
**Pull request by ancaantochi**
Removed a tests which tested the same actions as another test that was left.
Test was flaky because sometimes the c2d message was sent in the same time the device was disconnected and sometimes edgeHub received it and immediately the device proxy was not available. Fixed to check the device in IotHub that the device was disconnected before sending C2D message.

### CODE REVIEW

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs [hunk 1]
```diff
@@ -128,8 +85,16 @@ public async void Receive_C2D_SingleMessage_AfterOfflineMessage_ShouldSucceed()
                 // Dummy ReceiveAsync to ensure mqtt subscription registration before SendAsync() is called on service client.
                 await deviceClient.ReceiveAsync(TimeSpan.FromSeconds(1));
                 await deviceClient.CloseAsync();
-                // wait for the connection to be closed on the Edge side
-                await Task.Delay(TimeSpan.FromSeconds(30));
+
+                // Wait for the connection to be closed on the Edge side by checking device connection state
+                Task timerTask = Task.Delay(TimeSpan.FromSeconds(60));
```
Comment (nyanzebra) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:90]: Why bump to 60 seconds?

Comment (nyanzebra) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs]: Guess this is now bumped to 4 minutes? Will this be enough time to ensure the test is not flaky?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs [hunk 2]
```diff
@@ -247,7 +222,14 @@ async Task Cleanup(DeviceClient deviceClient, ServiceClient serviceClient, Regis
...
             if (deviceClient != null)
             {
-                await deviceClient.CloseAsync();
+                try
+                {
+                    await deviceClient.CloseAsync();
+                }
+                catch
+                {
+                    // ignore
```
Comment (nyanzebra) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:231]: Why would this fail? Is it worthwhile to just catch the specific exception(s) or do we just not care after cleanup call? (not saying it should be done differently, just trying to understand :) )

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs [hunk 3]
```diff
@@ -128,8 +85,16 @@ public async void Receive_C2D_SingleMessage_AfterOfflineMessage_ShouldSucceed()
                 // Dummy ReceiveAsync to ensure mqtt subscription registration before SendAsync() is called on service client.
                 await deviceClient.ReceiveAsync(TimeSpan.FromSeconds(1));
                 await deviceClient.CloseAsync();
-                // wait for the connection to be closed on the Edge side
-                await Task.Delay(TimeSpan.FromSeconds(30));
+
+                // Wait for the connection to be closed on the Edge side by checking device connection state
+                Task timerTask = Task.Delay(TimeSpan.FromMinutes(4));
```
Comment (vadim-kovalyov) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:90]: 4 minutes? I thought we are trying to speed our tests up, not slow them down :) On a serious note though, why do we need such long timeout?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs [hunk 4]
```diff
@@ -204,6 +169,16 @@ ITransportSettings[] GetTransportSettings(TransportType transportType = Transpor
             return settings;
         }
 
+        async Task WaitForDeviceDisconnected(RegistryManager rm, string deviceName)
+        {
+            var device = await rm.GetDeviceAsync(deviceName);
+            while (device.ConnectionState != DeviceConnectionState.Disconnected)
+            {
+                device = await rm.GetDeviceAsync(deviceName);
+                await Task.Delay(TimeSpan.FromSeconds(5));
```
Comment (vadim-kovalyov) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:178]: Why do we need it here? Is there a way to do it without delay?

**Review by vadim-kovalyov — COMMENTED** [review]
As a general rule, we should try to avoid long timeouts in tests, and ideally avoid having dependency on IotHub. Should we instead promote this to E2E test?

Also, after looking at this test, I'm trying to evaluate it's usefulness. So, we create device, connect, then disconnect, then try to send offline message and verify we received it.... It seems to me that we are testing Iot Hub, and not Iot Edge... So, do we really need this test?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs (hunk 1 above)
Comment (ancaantochi) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs]: I thought 60s would help, but it was still failing so just trying to see if 4min would help, but it was still failing. I found the issue and I think with the current change is stable, still running a few more times. So it needs to wait for connect before it disconnects. I opened a bug for the root cause issue, it is a corner case if device connects and quickly disconnects before edgeHub has time to connect to cloud, then cloud connection is not closed and c2d messages are sent and edgehub can't send it to device since it was disconnected.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs (hunk 3 above)
Comment (ancaantochi) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs]: I thought 60s would help, but it was still failing so just trying to see if 4min would help and it was still failing. I found the issue and I think with the current change is stable, still running a few more times. It needs to wait for connect before it disconnects. We are trying to fix it, not speed it :) Anyway I changed it and it waits up to 60s to connects and then disconnect and as soon this happens it can proceed, so it most cases it should be faster than 60s.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs (hunk 4 above)
Comment (ancaantochi) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:189]: I think there should be a delay so we don't make the calls to IotHub continuously. I picked 5s, it could be changed to lower value, maybe 3s?

**Comment by ancaantochi** [comment]
> As a general rule, we should try to avoid long timeouts in tests, and ideally avoid having dependency on IotHub. Should we instead promote this to E2E test? 
> Also, after looking at this test, I'm trying to evaluate it's usefulness. So, we create device, connect, then disconnect, then try to send offline message and verify we received it.... It seems to me that we are testing Iot Hub, and not Iot Edge... So, do we really need this test?

It is E2E test, just that it's not running in containers. I think it's useful because I found a bug for the root cause issue, it is a corner case if device connects and quickly disconnects before edgeHub has time to connect to cloud, then cloud connection is not closed, c2d messages are sent and edgehub can't send it to device because it was disconnected. I think long haul run into it before, with CI it's easy to debug from VS and actually find the root cause.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/TwinDiffE2ETest.cs [hunk 5]
```diff
@@ -377,6 +377,26 @@ static async Task<(DeviceClient, string)> InitializeDeviceClient(RegistryManager
...
+                try
+                {
+                    var device = await rm.GetDeviceAsync(deviceName);
+                    if (device.ConnectionState != DeviceConnectionState.Connected)
+                        throw new Exception("Device not connected to cloud");
+                    break;
+                }
+                catch (Exception)
+                {
+                    if (i == 4)
```
Comment (nyanzebra) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/TwinDiffE2ETest.cs:391]: nit: would you mind adding a small comment on this part for `if (i == 4)` ?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs (hunk 2 above)
Comment (ancaantochi) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:242]: it can fail from sdk if it was already closed, I would say we should not fail the test if close fails.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs [hunk 6]
```diff
@@ -224,7 +210,7 @@ Message CreateMessage(out string payload)
         [AssertionMethod]
         async Task VerifyReceivedC2DMessage(DeviceClient deviceClient, string payload, string p1Value)
         {
-            Client.Message receivedMessage = await deviceClient.ReceiveAsync();
+            Client.Message receivedMessage = await deviceClient.ReceiveAsync(TimeSpan.FromSeconds(20));
```
Comment (ancaantochi) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:213]: remove timeout, left by mistake

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/TwinDiffE2ETest.cs [hunk 7]
```diff
@@ -377,6 +377,26 @@ static async Task<(DeviceClient, string)> InitializeDeviceClient(RegistryManager
...
+                    {
+                        throw;
+                    }
+
+                    await Task.Delay(TimeSpan.FromSeconds(5));
```
Comment (nyanzebra) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/TwinDiffE2ETest.cs:396]: Is this for the same reason as the delay of 5s that Vadim commented on? Just to not call iothub a lot?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs (hunk 4 above)
Comment (nyanzebra) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:189]: long term planning question: Could these tests accomplish their goals if we had a mock iothub and we didn't have to worry about network calls? We just make some situation or set of events that should occur in the mock iothub?

If so, then I think having some delays here in these tests are fine. If not, then I like the idea of trying to have as small a delay as possible. :)

Comment (vadim-kovalyov) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:189]: I think we need to ask IotHub on their SLA for operations. 5sec is too much IMO. If I could, I'd probably do it with interval not more that 1s, and total timeout 10s max.

**Review by vadim-kovalyov — APPROVED** [review]
I don't want to block this, but I do recommend to still revisit the timeout numbers after we fix the root cause for the issue.

**Review by nyanzebra — APPROVED** [review]

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs (hunk 4 above)
Comment (ancaantochi) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/Cloud2DeviceTest.cs:189]: we can mock the sdk, so it won't connect to IotHub, but probably this issue wouldn't be caught. We should discuss in the tests improvements plan what should be mocked.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/TwinDiffE2ETest.cs (hunk 7 above)
Comment (ancaantochi) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/TwinDiffE2ETest.cs:396]: yes, it checks that device was connected to cloud
